packages/loopring_v3/contracts/impl/libtransactions/OwnerChangeTransaction.sol

```diff
@@ -0,0 +1,189 @@
+    );
+
+    /*event ChangeOwnerConsumed(
+        address  indexed owner,
+        address          newOwner
+    );*/
+
+    struct OwnerChange
+    {
+        address owner;
+        uint24  accountID;
+        uint16  feeTokenID;
+        uint    fee;
+        address newOwner;
+        uint32  nonce;
+        bytes32 walletHash;
+    }
+
+    // Auxiliary data for each owner change
+    struct OwnerChangeAuxiliaryData
+    {
+        bytes   signatureOldOwner;
+        bytes   signatureNewOwner;
+
+        address walletAddress;
+        bytes32 walletDataHash;
+        bytes   walletCalldata;
+    }
+
+    function process(
+        ExchangeData.State storage S,
+        ExchangeData.BlockContext memory ctx,
+        bytes memory data,
+        bytes memory auxiliaryData
+        )
+        internal
+        returns (uint /*feeETH*/)
+    {
+        OwnerChange memory accountTransfer = readOwnerChange(data);
+
+        OwnerChangeAuxiliaryData memory auxData = abi.decode(auxiliaryData, (OwnerChangeAuxiliaryData));
+
+        // Calculate the tx hash
+        bytes32 txHash = EIP712.hashPacked(
+            ctx.DOMAIN_SEPARATOR,
+            keccak256(
+                abi.encode(
+                    OWNERCHANGE_TYPEHASH,
+                    accountTransfer.owner,
+                    accountTransfer.accountID,
+                    accountTransfer.feeTokenID,
+                    accountTransfer.fee,
+                    accountTransfer.newOwner,
+                    accountTransfer.nonce,
+                    auxData.walletAddress,
+                    auxData.walletDataHash,
+                    keccak256(auxData.walletCalldata)
+                )
+            )
+        );
+
+        // Verify the signature if one is provided, otherwise fall back to an approved tx
+        if (auxData.signatureNewOwner.length > 0) {
+            require(txHash.verifySignature(accountTransfer.newOwner, auxData.signatureNewOwner), "INVALID_SIGNATURE");
+        } else {
+            require(S.approvedTx[accountTransfer.newOwner][txHash], "TX_NOT_APPROVED");
+            S.approvedTx[accountTransfer.newOwner][txHash] = false;
+        }
+
+        if (auxData.walletAddress == address(0)) {
+            // We allow the owner of the account to change when authorized by the current owner.
+            // Verify the signature if one is provided, otherwise fall back to an approved tx
+            if (auxData.signatureOldOwner.length > 0) {
+                require(txHash.verifySignature(accountTransfer.owner, auxData.signatureOldOwner), "INVALID_SIGNATURE");
+            } else {
+                require(S.approvedTx[accountTransfer.owner][txHash], "TX_NOT_APPROVED");
+                S.approvedTx[accountTransfer.owner][txHash] = false;
+            }
+        } else {
+            // If the account has a wallet, use it to recover the account
+            require(accountTransfer.walletHash != 0, "ACCOUNT_HAS_NO_WALLET");
```

Comment:
Currently it's only possible to an ACCOUNT_UPDATE with just an EdDSA signature, all other transactions need an ECDSA signature from the owner. This makes it very cheap to change keys and update the wallet data because we don't have to do anything onchain.

There are some edge cases with that actually for the wallet data. With just the EdDSA signature it would be possible to change the account owner, which in some scenario's would not be expected. But this is easily solved by having the owner actually sign the walletData (instead of just using the hash), so the operator needs to provide a signature from the old owner as well for using that data. This way we can keep this as cheap as possible as well.